### DISCUSSION TITLE
Running Audiobookshelf as a Windows Service

### DISCUSSION THREAD

**Discussion by petenun** [mikiher/audiobookshelf-windows#31] [General]
For anyone interested, I have been successfully running Audiobookshelf as a Windows service for several months and it has proven to be very stable in this configuration. Since the primary means of interacting with the application is through the Web UI, there is no sacrifice in functionality. The only change in behavior is, when installing an update, you must manually stop the service, run the installer, and then restart the service when completed. I've documented the procedure below in case anyone is interested.

**Please note: Though fully functional, this is not the way the author intended the application to be run. By running as a service, you are taking responsibility and should not expect the author (mikiher) to support this configuration.**

**What is a Windows Service and Why run Audio Bookshelf as a Windows service?**
In a nutshell, a windows service is a program that runs automatically when Windows starts. This differs from putting a program in the Startup folder in that services do not require a user to log-in in order to start and do not require a user to stay logged-in to continue running. Even if the computer is rebooted, the service will start up as soon as the machine comes back up. This is very handy when running on a home server that you want to be constantly available. In my case, I consider Audiobookshelf a critical media service that I want running 24x7 so there is always access to my audiobooks.

**What you'll need:**
The current Audiobookshelf-Windows installer
A program like NSSM (The Non-Sucking Service Manager) or similar (free download)
A computer on which to install Audiobookshelf.
Optional: A second computer on which your audiobooks are stored

**Instructions:**

1. Download and install the latest Audiobookshelf-Windows version, making note of where the executable files are saved. This is the same procedure you would use whether installing as a service or not. This would also be the time to set the ABS port as needed, and you can even configure the program. The important part is that it's installed and functional. 

2. When you're ready to transition to a Windows Service, exit the program.

3. Install and launch NSSM. (The Windows SC command can also be used, as well as other 3rd party tools).  In the case of NSSM, open a command prompt (not powershell) as Administrator and type "nssm install". This will bring up the following dialogue.

![image](https://github.com/user-attachments/assets/d2bb7339-1d3b-47f4-ac98-8807f0e699ba)

4. On the Application tab, under Path, browse to the location of your audiobookshelf.exe file. The Startup Directory should auto-populate with this same location.  Be sure to specify the Service Name at the bottom. This is how the service will appear in the Services MMC.

5. [Optional] In the event that your audiobook files are located on a different computer, you will need to ensure that the serive is running with an account that has access to the files on that server. Switch to the "Log on" tab of the NSSM installer and specify a different account that has access. If you are in a Windows Workgroup, I suggest creating an identical account on both computers with the same name and password, then use that for access. If you are running in an Active Directory domain environment, you can just use the domain account and password.

![image](https://github.com/user-attachments/assets/6d357fef-4192-4066-9a7b-e03c1841107e)

6. Click on Install Service. The service should be created. Keep in mind that it won't be running until you start it.

7. Open Services. The easiest way is to click Start and type "Services", or you can type "servcies.msc". Once Services comes up, verify that you see the Audiobookshelf service and that its Starup Type is set to Automatic. Also verify the "Log On As" is set to your account if you changed it in Step 5.

![image](https://github.com/user-attachments/assets/d9ca15d4-d219-45d0-9ac9-658d194ee34c)

8. Start the service. Just right-click on the service and select Start. You should see the Status switch ro "Running". At this point Audiobookshelf will be running as a service.

9. Open a web browser and verify that you can connect to Audiobookshelf like you normally would. The program will function as it always has, with configurations saved to the same location. Nothing should be different.

 **** Important Note** when Updating Audiobookshelf. Before running the installer for a new version of AudioBookshelf, you must manually stop the service before installing, and then restart the service after a successful installation. Do not have the installer run the program after installation, as it will run in the foreground, not as a service. If you start the service in this state, you will have two instances running and there will likely be conflicts if they run at all. 

I hope this has been helpful.

**Comment by mikiher** [comment]
Very nice. I have considered implementing ABS as a service, but in the end I decided to make it as simple as possible for the sake of not complicating things for novice users. 

If there's a lot of interest in making this (installing as a service) a feature of audiobokshelf-windows, I'm willing to consider it.

**Reply by petenun** [reply]
Honestly, I think you took the best approach to make the program usable for people of all skill levels. It's better to spread the love to more users rather than just the techno-savvy ones. I'm just glad that it does work so well as a service! 😀 

In case it needs to be said, thanks for putting your love and time into this project!!

**Reply by kanine** [reply]
> in the end I decided to make it as simple as possible for the sake of not complicating things for novice users.

I run a health check on various servers on a regular basis and restart them when they crash. Particularly Plex! Anyways it works by polling the various ports for a response and if dead it will restart the process again using command line / api or whatever means makes sense for the particular app (usually a Task Scheduler process for maximum compatibility).

Is there a way from the command line to start / stop the server with a command line option on the exe? My alternative approach is going to be to kill the exe if it's running and just restart but command line start/stop server would be more elegant.

**Reply by petenun** [reply]
I'll leave it to mikher to discuss command line switches.

In my case, since I'm running as a service, it's a simple "net stop/start audiobookshelf" from a command prompt.

**Reply by mikiher** [reply]
There's no way to do this with a command line option. If you want a workaround, @petenun's will likely work.

**Reply by kanine** [reply]
At this point it's a solution looking for a problem hopefully the server proves to be stable I just switched from docker yesterday.

**Reply by petenun** [reply]
It very well may be a moot point. I've been running ABS on Windows since April as a 24x7 service and it has been rock solid. The only restarts I perform are when there are updates.

**Reply by kanine** [reply]
Sounds good, I couldn't help myself though haha. Got it checking nginx for the reverse proxy and also the ABS Port + tray.exe. Works but doubt it will ever trigger unlike some of the other checks I had to put in place for less stable apps. Thanks for all the info, and making this possible in the first place.

**Comment by binyaminyblatt** [comment]
I don’t have all my audiobooks locally most are stored on network shares how do I make sure that I can still access them?

**Comment by petenun** [comment]
I'd be happy to help if I can, but I need more information.

Network Share
- Is your network share on your local LAN with at least gigabit wired network connectivity?
- What OS is your network share running? Windows, Linux, NAS (if so which one)?
- Is your share available 24/7 (always on)?
- Is it an option to run ABS on the same machine on which your files are stored?

Computer (the machine that is running ABS)
- It's running Windows, I presume? If so, which version?
- Is it on 24/7
- Does it have (at least) a wired gigabit network connection?
- Is it always on the same LAN as the network share, or is it a laptop that can move around?

AudioBookshelf
- How are you accessing (or do you want to access) AudioBookshelf? Web browser? Mobile app?
- Do you want to use it exclusively on your LAN or from outside your LAN via Internet?

**Comment by binyaminyblatt** [comment]
I did some experimenting and everything is working

**Comment by binyaminyblatt** [comment]
It's so much more stable as a service then with the starter app

**Reply by binyaminyblatt** [reply]
Though I wish the command line switches where better documented
